libraries/abstractions/secure_sockets/utest/secure_sockets_utest.c

```diff
@@ -0,0 +1,1334 @@
+    TEST_ASSERT_EQUAL_INT( SOCKETS_ERROR_NONE, ret );
+}
+
+static Socket_t create_TLS_connection()
+{
+    Socket_t s = initSocket();
+
+    setTLSMode( s );
+    TLS_Init_IgnoreAndReturn( pdFREERTOS_ERRNO_NONE );
+    TLS_Connect_IgnoreAndReturn( pdFREERTOS_ERRNO_NONE );
+    connectSocket( s );
+    return s;
+}
+
+static Socket_t create_normal_connection()
+{
+    Socket_t s = initSocket();
+
+    connectSocket( s );
+    return s;
+}
+/* ======================  TESTING SOCKETS_Send  ============================ */
+
+/*!
+ * @brief A happy send case with normal sockets
+ */
+void test01_SecureSockets_send_successful( void )
+{
+    int32_t ret;
+    Socket_t so = create_normal_connection();
+    const char buffer[ BUFFER_LEN ];
+
+    lwip_send_ExpectAnyArgsAndReturn( BUFFER_LEN );
+    ret = SOCKETS_Send( so,
+                        buffer,
+                        BUFFER_LEN,
+                        0 );
+    TEST_ASSERT_EQUAL_INT( BUFFER_LEN, ret );
+    deinitSocket( so );
+}
+
+/*!
+ * @brief A happy send case with tls sockets
+ */
+void test02_SecureSockets_send_successful_tls( void )
+{
+    int32_t ret;
+    const char buffer[ BUFFER_LEN ];
+
+    Socket_t so = create_TLS_connection();
+
+    TLS_Send_ExpectAnyArgsAndReturn( BUFFER_LEN );
+    ret = SOCKETS_Send( so,
+                        buffer,
+                        BUFFER_LEN,
+                        0 );
+    TEST_ASSERT_EQUAL_INT( BUFFER_LEN, ret );
+    TLS_Cleanup_ExpectAnyArgs();
+    deinitSocket( so );
+}
+
+/*!
+ * @brief Test various bad parameters
+ */
+void test03_SecureSockets_send_invalid_parameters( void )
+{
+    Socket_t s = SOCKETS_INVALID_SOCKET;
+    const char buffer[ BUFFER_LEN ];
+    int32_t ret;
+
+    /* test invalid socket */
+    ret = SOCKETS_Send( s,
+                        buffer,
+                        BUFFER_LEN,
+                        0 );
+    TEST_ASSERT_EQUAL_INT( SOCKETS_SOCKET_ERROR, ret );
+
+    /* test null buffer */
+    s = initSocket();
+    ret = SOCKETS_Send( s,
+                        NULL, /* null buffer */
+                        BUFFER_LEN,
+                        0 );
+    TEST_ASSERT_EQUAL_INT( SOCKETS_EINVAL, ret );
+
+    /* test zero datalength */
+    ret = SOCKETS_Send( s,
+                        buffer,
+                        0,
+                        0 );
+    TEST_ASSERT_EQUAL_INT( SOCKETS_EINVAL, ret );
+    deinitSocket( s );
+}
+
+void test04_SecureSockets_send_not_connected( void )
+{
+    int32_t ret;
+    const char buffer[ BUFFER_LEN ];
+
+    Socket_t so = initSocket();
+
+    ret = SOCKETS_Send( so,
+                        buffer,
+                        BUFFER_LEN,
+                        0 );
+    TEST_ASSERT_EQUAL_INT( SOCKETS_ENOTCONN, ret );
+    deinitSocket( so );
+}
+
+/* =====================  TESTING SOCKETS_Socket  =========================== */
+
+/*!
+ * @brief calling SOCKETS_Socket with various invalid arguments
+ *        expects SOCKETS_INVALID_SOCKET
+ */
```

Comment:
In general I would like to see us increase the verbosity of the description of the test. Specifically I would like to explicitly have a section for every test that describes "Purpose of this test"

Reply:
I will try to find an appropriate doxygen tag for that and add more verbose descriptions